Update documentation about restoring from backup (#1678)
scylladb/scylla-operator: docs/source/nodeoperations/restore.md

This procedure will describe how to restore from backup taken using [Scylla Manager](../manager.md) to a fresh **empty** cluster of any size.

:::{caution}
Due to a [bug](https://github.com/scylladb/scylla-manager/issues/3679) in Scylla Manager not supporting ScyllaClusters having both non-TLS and TLS CQL ports open, you have to disable the TLS certificate management
in Scylla Operator.
This step will no longer be required, when the bug is fixed.

To disable TLS certificate management in Scylla Operator add `--feature-gates="AutomaticTLSCertificates=false"` flag to Scylla Operator deployment.

```console

> kubectl -n scylla-operator patch deployment/scylla-operator --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value": "--feature-gates="AutomaticTLSCertificates=false"}]'

Reviewer: This failed for me - the doublequotes are invalid.
```suggestion
kubectl -n scylla-operator patch deployment/scylla-operator --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value": "--feature-gates=AutomaticTLSCertificates=false"}]'
```

Author: Thanks, fixed

This procedure will describe how to restore from backup taken using [Scylla Manager](../manager.md) to a fresh **empty** cluster of any size.

:::{caution}

> Due to a [bug](https://github.com/scylladb/scylla-manager/issues/3679) in Scylla Manager not supporting ScyllaClusters having both non-TLS and TLS CQL ports open, we have to disable TLS certificate management

Reviewer: ```suggestion
Due to a [bug](https://github.com/scylladb/scylla-manager/issues/3679) in Scylla Manager not supporting ScyllaClusters having both non-TLS and TLS CQL ports open, you have to disable TLS certificate management
```

Author: done

> in Scylla Operator before you restore your cluster. Once bug is fixed, this is no longer required.

Reviewer: ```suggestion
This step will no longer be required, when the bug is fixed.
```

Author: done

> To disable TLS certificate management in Scylla Operator add `--feature-gates="AutomaticTLSCertificates=false"` to Scylla Operator deployment.

Reviewer: ```suggestion
To disable the TLS certificate management in Scylla Operator add `--feature-gates="AutomaticTLSCertificates=false"` flag to Scylla Operator deployment.
```

Author: done

```console
kubectl -n scylla-operator patch deployment/scylla-operator --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value": "--feature-gates="AutomaticTLSCertificates=false"}]'
```

:::

> In this example our ScyllaCluster from which we took backup is called `source`. Backup will be restored into the ScyllaCluster called `target`.

Reviewer: ```suggestion
In the following example, the ScyllaCluster, which was used to take the backup, is called `source`. We will be restoring the backup into the ScyllaCluster named `target`.
```

Author: done

::::{tab-set}
:::{tab-item} Source ScyllaCluster
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
metadata:
  name: source
  namespace: scylla
spec:
  agentVersion: 3.2.5
  version: 5.4.1
  developerMode: true
  backups:
  - name: foo
    location:
    - s3:source-backup
    keyspace:
    - '*'
  datacenter:
    name: us-east-1
    racks:
    - name: us-east-1a
      members: 1
      storage:
        capacity: 1Gi
      resources:
        limits:
          cpu: 1
          memory: 1Gi
```
:::
:::{tab-item} Target ScyllaCluster
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
metadata:
  name: target
  namespace: scylla
spec:
  agentVersion: 3.2.5
  version: 5.4.1
  developerMode: true
  datacenter:
    name: us-east-1
    racks:
    - name: us-east-1a
      members: 1
      storage:
        capacity: 1Gi
      resources:
        limits:
          cpu: 1
          memory: 1Gi
```
:::
::::

> Make sure our target cluster is already registered in Scylla Manager, to get list of all registered clusters execute following command in Scylla Manager Pod:

Reviewer: ```suggestion
First, make sure your target cluster is already registered in Scylla Manager. To get a list of all registered clusters, execute the following command in the Scylla Manager Pod:
```

Reviewer: On a second thought though, I wouldn't say `run the following command in X`, but post a `kubectl exec` command against the manager deployment in each of those steps, .i.e. `kubectl exec -n scylla-manager -it deployment.apps/scylla-manager -- sctool cluster list`. I think it's more bulletproof.

Author: done

Reviewer: Some weird hybrid came out of that now: `Make `target` cluster is already registered in Scylla Manager`

```console
$ sctool cluster list
+--------------------------------------+---------------------------------------+---------+-----------------+
| ID                                   | Name                                  | Port    | CQL credentials |
+--------------------------------------+---------------------------------------+---------+-----------------+
| af1dd5cd-0406-4974-949f-dc9842980080 | scylla/target                        | default | set             |
| ebd82268-efb7-407e-a540-3619ae053778 | scylla/source                        | default | set             |
+--------------------------------------+---------------------------------------+---------+-----------------+
```

> Identify to which snapshot you want to restore. To get list of available snapshot execute following command on Scylla Manager Pod.

Reviewer: ```suggestion
Identify the tag of a snapshot which you want to restore. To get a list of all available snapshots, execute the following command in the Scylla Manager Pod.
```

Reviewer: ditto kubectl exec

Author: done

```console
sctool backup list -c <CLUSTER_ID> --all-clusters -L <BACKUP_LOCATION>
```

Where:

> * `CLUSTER_ID` - is a name or ID of registered cluster having access to `BACKUP_LOCATION`. 

Reviewer: ```suggestion
* `CLUSTER_ID` - the name or ID of a registered cluster with access to `BACKUP_LOCATION`. 
```

Author: done

> * `BACKUP_LOCATION` - is a location where backup is stored.

Reviewer: ```suggestion
* `BACKUP_LOCATION` - the location in which the backup is stored.
```

Author: done

> In our example `BACKUP_LOCATION` is `s3:source-backup` and we will use `scylla/target` cluster as `CLUSTER_ID` to make sure it has access to bucket.

Reviewer: ```suggestion
In this example, `BACKUP_LOCATION` is `s3:source-backup`. Use the name of cluster which has access to the backup location for `CLUSTER_ID`. In this example, it's `scylla/target`.
```

```console

> $ sctool backup list -c scylla/target --all-clusters -L s3:source-backup

Reviewer: ditto kubectl exec

Author: done

backup/ff36d7e0-af2e-458c-afe6-868e0f3396b2
Snapshots:
  - sm_20240105115931UTC (409MiB, 1 nodes)
Keyspaces:
  - system_schema (15 tables)
  - users (9 tables)

```

> We are going to restore the cluster using `sm_20240105115931UTC` snapshot.

Reviewer: ```suggestion
In the below commands, we are restoring the `sm_20240105115931UTC` snapshot. Replace it with a tag of a snapshot that you want to restore.
```

Author: done

> Restoring consist of two steps, first we restore the schema, and then the data.

Reviewer: ```suggestion
Restoring consist of two steps. First, you'll restore the schema, and then the data.
```

Author: done

> To restore schema, create a restore task manually on target ScyllaCluster by executing following command in Scylla Manager Pod.

Reviewer: ```suggestion
To restore schema, create a restore task manually on target ScyllaCluster by executing the following command in the Scylla Manager Pod.
```

Author: done

```console

> sctool restore -c <CLUSTER_ID> -L <BACKUP_LOCATION> -T <SNAPSHOT_TAG> --restore-schema

Reviewer: ditto kubectl exec

Author: done

```

Where:

> * `CLUSTER_ID` - is a name or ID of a cluster we are restoring to.

Reviewer: ```suggestion
* `CLUSTER_ID` - a name or ID of a cluster you want to restore into.
```

Author: done

> * `BACKUP_LOCATION` - location where backup is stored.

Reviewer: ditto

Author: done

> * `SNAPSHOT_TAG` - snapshot tag of a backup we are restoring from.

Reviewer: ```suggestion
* `SNAPSHOT_TAG` - a tag of a snapshot that you want to restore.
```

Author: done

> After task is created, output of the command will return ID of a restore task.

Reviewer: ```suggestion
When the task is created, the command will output the ID of a restore task.
```

Author: done

```console
$ sctool restore -c scylla/target -L s3:source-backup -T sm_20240105115931UTC --restore-schema
restore/57228c52-7cf6-4271-8c8d-d446ff160747
```

> To check progress use:

Reviewer: ```suggestion
Use the following command to check progress of the restore task:
```

Author: done

```console
$ sctool progress -c scylla/target restore/57228c52-7cf6-4271-8c8d-d446ff160747

> Restore progress

Reviewer: can we hide the outputs? Iirc the plugin we're using allows to display it on-click and hide by default.

Author: I tried, with admonition having toggle class but it didn't work. It looked bad, and toggle was missing. Haven't spend more time on it so leaving as is.

Run:            0dd20cdf-abc4-11ee-951c-6e7993cf42ed
Status:         DONE - restart required (see restore docs)
Start time:     05 Jan 24 12:15:02 UTC
End time:       05 Jan 24 12:15:09 UTC
Duration:       6s
Progress:       100% | 100%
Snapshot Tag:   sm_20240105115931UTC

+---------------+-------------+----------+----------+------------+--------+
| Keyspace      |    Progress |     Size |  Success | Downloaded | Failed |
+---------------+-------------+----------+----------+------------+--------+
| system_schema | 100% | 100% | 214.150k | 214.150k |   214.150k |      0 |
+---------------+-------------+----------+----------+------------+--------+
```

> As suggested in the progress output, we need to execute a rolling restart of ScyllaCluster.

Reviewer: ```suggestion
As suggested in the progress output, you will need to execute a rolling restart of the ScyllaCluster.
```

Author: done

```console
kubectl patch scyllacluster target --type merge -p '{"spec": {"forceRedeploymentReason": "schema restored"}}'
```

> And wait until it's finished:

Reviewer: ditto (use the below commands)

Author: done

```console
$ kubectl wait --for='condition=Progressing=False' -n scylla scyllaclusters.scylla.scylladb.com/target
scyllacluster.scylla.scylladb.com/target condition met

$ kubectl wait --for='condition=Degraded=False' -n scylla scyllaclusters.scylla.scylladb.com/target
scyllacluster.scylla.scylladb.com/target condition met

$ kubectl wait --for='condition=Available=True' -n scylla scyllaclusters.scylla.scylladb.com/target
scyllacluster.scylla.scylladb.com/target condition met
```

> Once completed, we can proceed to restoring the data.

Reviewer: ```suggestion
Proceed to restoring the data when it's completed.
```
or I would just delete this

Author: removed

> To restore tables content, create a restore task manually on target ScyllaCluster by executing following command in Scylla Manager Pod.

Reviewer: ```suggestion
To restore the tables' content, create a restore task manually on target ScyllaCluster by executing the following command in the Scylla Manager Pod.
```

Author: done

```console

> sctool restore -c <CLUSTER_ID> -L <BACKUP_LOCATION> -T <SNAPSHOT_TAG> --restore-tables

Reviewer: ditto `kubectl exec`

Reviewer: you're using variables in some of the commands and set values in the others, let's make this consistent

Author: added kubectl, but I don't get what you mean in the second comment. Having command template and then example is in my view beneficial to understanding what needs to be changed.

Reviewer: > Having command template and then example is in my view beneficial to understanding what needs to be changed.

my bad, didn't notice that it's only for the ones with an example output

```

Where:

> * `CLUSTER_ID` - is a name or ID of a cluster we are restoring to.

Reviewer: ditto

Author: done

> * `BACKUP_LOCATION` - location where backup is stored.

Reviewer: ditto

Author: done

> * `SNAPSHOT_TAG` - snapshot tag of a backup we are restoring from.

Reviewer: ditto

Author: done

> After task is created, output of the command will return ID of a restore task.

Reviewer: ditto

Author: done

```console

> $ sctool restore -c scylla/target -L s3:source-backup -T sm_20240105115931UTC --restore-tables

Reviewer: ditto kubectl exec

Author: done

restore/63642069-bed5-4def-ba0f-68c49e47ace1
```

> To check progress use:

Reviewer: ditto

```console
$ sctool progress -c scylla/target restore/63642069-bed5-4def-ba0f-68c49e47ace1

> Restore progress

Reviewer: ditto (about output)

Run:            ab015cef-abc8-11ee-9521-6e7993cf42ed
Status:         DONE
Start time:     05 Jan 24 12:48:04 UTC
End time:       05 Jan 24 12:48:15 UTC
Duration:       11s
Progress:       100% | 100%
Snapshot Tag:   sm_20240105115931UTC

+-------------+-------------+--------+---------+------------+--------+
| Keyspace    |    Progress |   Size | Success | Downloaded | Failed |
+-------------+-------------+--------+---------+------------+--------+
| users       | 100% | 100% | 409MiB |  409MiB |     409MiB |      0 |
+-------------+-------------+--------+---------+------------+--------+

Post-restore repair progress
Run:            ab015cef-abc8-11ee-9521-6e7993cf42ed
Status:         DONE
Start time:     05 Jan 24 12:48:04 UTC
End time:       05 Jan 24 12:48:15 UTC
Duration:       11s
Progress:       100%
Intensity:      1
Parallel:       0
Datacenters:
  - us-east-1

+-------------+--------------+----------+----------+
| Keyspace    |        Table | Progress | Duration |
+-------------+--------------+----------+----------+
| users       | users        | 100%     | 0s       |
+-------------+--------------+----------+----------+

```

> As last step, we need to manually reset `tombstone_gc` via `cqlsh` in target ScyllaCluster on **every table and every keyspace** which was restored.

Reviewer: ```suggestion
When the restore procedure has finished, you will need to manually reset `tombstone_gc` via `cqlsh` in target ScyllaCluster on **every table and every keyspace** which was restored.
```

Reviewer: "last step" is not going to age well if we ever extend it

Author: done

> As last step, we need to manually reset `tombstone_gc` via `cqlsh` in target ScyllaCluster on **every table and every keyspace** which was restored.

Reviewer 2: Looking at https://manager.docs.scylladb.com/stable/restore/restore-tables.html

> Restore procedure works iteratively over restored locations, manifests and tables.
> - ...
> - Repair restored tables
> - Reset all restored tables tombstone_gc to its original mode

This made an impression on me that it's handled by the manager...

Reviewer: The examples https://manager.docs.scylladb.com/stable/restore/examples.html mention it as a manual step.
/cc @Michal-Leszczynski 
could you have a look at the procedure here?

Author: Looks like it's not in sync as following page explicitly mentions it needs to be done. Although it also suggests to run a repair which is run automatically.
https://manager.docs.scylladb.com/stable/restore/examples.html

@Michal-Leszczynski is this step to manually set tombstone_gc required or it's already automated in Manager?

Reviewer 3: Sorry for the confusion, with SM 3.2 there is no need for manual tombstone_gc alteration nor repair. Examples in the docs are outdated and they need to change.

Reviewer: thanks Michal, logged https://github.com/scylladb/scylla-manager/issues/3684

Author: removed this step
